File: 1122.relative-sort-array.cpp
/*
 * @lc app=leetcode id=1122 lang=cpp
 *
 * [1122] Relative Sort Array
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
    std::unordered_map<int, int> val2idx;
    for (int i = 0; i < arr2.size(); i++) {
      val2idx[arr2[i]] = i;
    }
    std::sort(arr1.begin(), arr1.end(), [&val2idx](auto lhs, auto rhs) {
      if (val2idx.count(lhs) == 0) {
        if (val2idx.count(rhs) == 0) {
          // ascending order
          return lhs < rhs;
        } else {
          // need to swap
          return false;
        }
      } else {
        if (val2idx.count(rhs) == 0) {
          // no need to swap
          return true;
        } else {
          return val2idx[lhs] < val2idx[rhs];
        }
      }
    });
    return arr1;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> arr1, arr2, exp;
  arr1 = {2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19};
  arr2 = {2, 1, 4, 3, 9, 6};
  exp = {2, 2, 2, 1, 4, 3, 3, 9, 6, 7, 19};
  ASSERT_EQ(s.relativeSortArray(arr1, arr2), exp);
  return 0;
}

File: 1122.relative-sort-array.rs
/*
 * @lc app=leetcode id=1122 lang=rust
 *
 * [1122] Relative Sort Array
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn relative_sort_array(mut arr1: Vec<i32>, arr2: Vec<i32>) -> Vec<i32> {
        let mut cnt = [0; 1021];
        for n in arr1.iter() {
            cnt[*n as usize] += 1;
        }
        let mut i = 0;
        for n in arr2 {
            while cnt[n as usize] > 0 {
                cnt[n as usize] -= 1;
                arr1[i] = n;
                i += 1;
            }
        }
        for (n, c) in cnt.iter_mut().enumerate() {
            while *c > 0 {
                *c -= 1;
                arr1[i] = n as i32;
                i += 1;
            }
        }
        arr1
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::relative_sort_array(
            vec![2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19],
            vec![2, 1, 4, 3, 9, 6]
        ),
        vec![2, 2, 2, 1, 4, 3, 3, 9, 6, 7, 19]
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::relative_sort_array(vec![28, 6, 22, 8, 44, 17], vec![22, 28, 8, 6]),
        vec![22, 28, 8, 6, 17, 44]
    );
}
